File: src/ruby_nmea/nmea.rb
require 'net/http'
require 'uri'
require 'csv'

def parse(line)
  line = line.strip
  tokens = line.split(',')
  return [-1, 0, 0, 0, 0, 0] unless tokens[0] == "$GPGGA"
  return [-3, 0, 0, 0, 0, 0] unless [9, 15].include?(tokens.length)

  if tokens.length == 15
    message_id, utc, lat, sn, lon, we, gps_quality, svs, hdop, height, height_unit, geoid_sep, geoid_sep_meters, age, station_id_ctrl = tokens
  else
    message_id, utc, lat, sn, lon, we, gps_quality, svs, station_id_ctrl = tokens
  end

  station_id, ctrl = station_id_ctrl.split('*')

  checksum = line[1..-4].chars.reduce(0) { |sum, ch| sum ^ ch.ord }
  checksum_str = checksum.to_s(16).upcase.rjust(2, '0')
  return [-2, 0, 0, 0, 0, 0, 0] if checksum_str != ctrl

  hour = utc[0, 2].to_f
  minute = utc[2, 2].to_f
  sec = utc[4..-1].to_f

  lat_dg = lat[0, 2].to_i
  lat_min = lat[2..-1].to_f
  latitude = lat_dg + lat_min / 60.0
  latitude *= -1 if sn == 'S'

  lon_dg = lon[0, 3].to_i
  lon_min = lon[3..-1].to_f
  longitude = lon_dg + lon_min / 60.0
  longitude *= -1 if we == 'W'

  [0, hour, minute, sec, latitude, longitude, height]
end

def parse_lines(lines)
  lines.map { |line| parse(line) }.select { |parsed| parsed[0] == 0 }
end

def parse_file(filename)
  parse_lines(File.readlines(filename))
end

def parse_bytes(filename)
  chars = CSV.read(filename)
  string = ''
  chars[0].each { |char| 
    val = char.to_i
    str = val.ord.chr
    string = string + str
  }

  parse_lines(string.lines)
end

def calculate_center(values)
  values.sum / values.size.to_f
end

def generate_static_map(filename, image_filename, api_key)
  lines = File.readlines(filename)
  parsed_data = parse_lines(lines)
  lats = parsed_data.map { |line| line[-3] }
  lons = parsed_data.map { |line| line[-2] }

  center_lat = calculate_center(lats)
  center_lon = calculate_center(lons)
  center = "#{center_lat},#{center_lon}"
  zoom = '13'
  size = '600x300'
  mtype = 'roadmap'
  markers = lats.each_with_index.map { |lat, i| "color:blue%7Clabel:#{i}%7C#{lat},#{lons[i]}" }
  
  uri = URI("https://maps.googleapis.com/maps/api/staticmap?center=#{center}&zoom=#{zoom}&size=#{size}&maptype=#{mtype}&key=#{api_key}")
  markers.each { |marker| uri.query += "&markers=#{marker}" }
  
  response = Net::HTTP.get_response(uri)
  File.open(image_filename, 'wb') { |file| file.write(response.body) } if response.is_a?(Net::HTTPSuccess)
end

def main
  parse_bytes('test_files/bytes.nmea').each { |position| puts position.inspect }
end

main if __FILE__ == $PROGRAM_NAME

File: src/python_nmea/nmea.py
import requests 
def parse(line):
    line = line.strip()
    tokens = line.split(',')
    if tokens[0] != "$GPGGA":
        return [-1,0,0,0,0,0]
    if (len(tokens)) != 15 and (len(tokens)) != 9:
        return [-3,0,0,0,0,0]
    if((len(tokens))) == 15:
        [messageId, UTC, Lat, SN, Lon, WE, GPS_quality, SVs, HDOP, height, height_unit, geoid_sep, geoid_sep_meters, age, stationID_ctrl] = tokens
    if((len(tokens))) == 9:
        [messageId, UTC, Lat, SN, Lon, WE, GPS_quality, SVs, stationID_ctrl] = tokens
    [stationID, ctrl] = stationID_ctrl.split('*')
   
    checksum = 0
    for ch in line[1:-3]:
        checksum = checksum ^ ord(ch)
    
    nibble1 = int(checksum / 16)
    if nibble1 == 10: nibble1 = 'A'
    if nibble1 == 11: nibble1 = 'B'
    if nibble1 == 12: nibble1 = 'C'
    if nibble1 == 13: nibble1 = 'D'
    if nibble1 == 14: nibble1 = 'E'
    if nibble1 == 15: nibble1 = 'F'
    nibble2 = checksum % 16
    if nibble2 == 10: nibble2 = 'A'
    if nibble2 == 11: nibble2 = 'B'
    if nibble2 == 12: nibble2 = 'C'
    if nibble2 == 13: nibble2 = 'D'
    if nibble2 == 14: nibble2 = 'E'
    if nibble2 == 15: nibble2 = 'F'
    checksum = str(nibble1) + str(nibble2)
    if(checksum != ctrl): 
        print(checksum, ctrl)
        return [-2,0,0,0,0,0,0] 


    hour = float(UTC[0:2])
    minute = float(UTC[2:4])
    sec = float(UTC[4:])

    lat_dg = int(Lat[0:2])
    lat_min = float(Lat[2:])
    lat = lat_dg + lat_min/60
    if SN == 'S':
        lat = lat * -1
    
    lon_dg = int(Lon[0:3])
    lon_min = float(Lon[3:])
    lon = lon_dg + lon_min/60
    if WE == 'W':
        lon *= -1
    
    
    
    

    return [0,hour, minute, sec, lat, lon, height]

def parseLines(lines):
    output = []
    for line in lines:
        parsed = parse(line)
        if parsed[0] == 0:
            output.append(parsed)
    return output

def parseFile(filename):
    lines = open(filename).readlines()
    return parseLines(lines)

def parseBytes(filename):
    chars = eval(open(filename).read())
    string = ''
    for ch in chars:
        string += chr(ch)
    return parseLines(string.split('\n'))
        

#The code below generates Google Static map for the provided set of lines. If you want to use it get the Google API key; keep in mind that static maps is a paid service (cost of a single image is less than 1 grosz, but still you need to connect your card and stuff).
api_key = 'XXXXXXXXXXXXXXXXX'

def calculateCenter(values):
    total = 0
    for value in values:
        total += value
    return total/(1.0*len(values))

def generateStaticMap(filename, image_filename):
    with open(filename) as file:
        lines = file.readlines()
        lats = []
        lons = []
        markers = []
        for parsedLine in parseLines(lines):
            lats.append(parsedLine[-3]) 
            lons.append(parsedLine[-2]) 
        centerLat = calculateCenter(lats)
        centerLon = calculateCenter(lons)
        center = f'{centerLat},{centerLon}'        
        zoom = '13'
        size = '600x300'
        mtype = 'roadmap'
        markers = []
        for i, _ in enumerate(lats):
            markers.append(f'color:blue%7Clabel:{i}%7C{lats[i]}, {lons[i]}')
        request = f"https://maps.googleapis.com/maps/api/staticmap?center={center}&zoom={zoom}&size={size}&maptype={mtype}"
        for marker in markers:
            request += f'&markers={marker}'
        request += f'&key={api_key}'
        response = requests.get(request)

        with open(image_filename,'wb+') as ofile:    
            ofile.write(response.content)
          
def main():
    for position in     parseBytes('test_files/bytes.nmea'):
        print(position)
if __name__ == '__main__':
    main()


